keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1543 @@
+make "fail closed" a safe mode of operation. As a concrete example, only
+allowing CEL expressions that pass compilation and type checking significantly
+reduces the opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare "zero trust" policies is
+important to compliance. By "zero trust", we mean policy rules that apply
+automatically to newly created resources (e.g. namespaces) in a restrictive way
+but that can be made less restrictive on a more granular level (e.g. per
+namespace).
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to webhook Rules, NamespaceSelector,
+  ObjectSelector)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  and
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Automatic version conversion support (similar to webhooks MatchPolicy=Equivalent), with ability to introspect GVK of admitted resource in CEL expression.
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+severity, levels, failure policies, type safety, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  constraints:
+    rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+    - expression: "object.spec.replicas <= config.maxReplicas"
```

Comment:
`expression` should have a `language` field.

e.g.

```yaml
validations:
- expression: "object.spec.replicas <= config.maxReplicas"
  language: cel
...
```

The reason for this is to support future additional languages that we may decide to add in the future. Implicit expression languages can be defaulted in, but will be present in the object represantion. So it doesn't hinder usability, but it does make the API object more future-proof/flexible.

Reply:
For CEL Validation Rules, we had a small section about additional language support: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#other-expression-languages

The idea was that when we introduce a second language, we add the field to allow the user to pick it. Since we always have a expression defined in a struct where the language field can be added later, we're future proofed.